src/platforms/hosted/bmp_libusb.c

```diff
@@ -47,291 +106,373 @@ void libusb_exit_function(bmp_info_s *info)
 				}
-				if (found) {
-					active_cable = cable->name;
-					type = BMP_TYPE_FTDI;
-					break;
+				if (manufacturer == NULL) {
+					manufacturer = get_device_descriptor_string(handle, device_descriptor->iManufacturer);
+				}
+				if (serial == NULL) {
+					//
+					// If this is an STLink V2 it does not report serial number correctly, for bypass
+					// trying to read it
+					//
+					if (device_descriptor->idVendor != VENDOR_ID_STLINK &&
+						device_descriptor->idProduct != PRODUCT_ID_STLINKV2)
+						serial = get_device_descriptor_string(handle, device_descriptor->iSerialNumber);
+					else
+						serial = strdup("---");
+				}
+
+				if (version == NULL) {
+					version = strdup("---");
 				}
+
+				*probe_list = probe_info_add_by_id(*probe_list, probe_type, device, device_descriptor->idVendor,
+					device_descriptor->idProduct, manufacturer, product, serial, version);
+				probe_added = true;
+				libusb_close(handle);
 			}
-			if (!cable->name)
-				continue;
+			break;
 		}
+		vid_pid_index++;
+	}
+	return probe_added;
+}
 
-		if (report)
-			DEBUG_WARN("%2zu: %s, %s, %s\n", found_debuggers + 1U, serial[0] ? serial : NO_SERIAL_NUMBER, manufacturer,
-				product);
-
-		info->libusb_dev = dev;
-		info->vid = desc.idVendor;
-		info->pid = desc.idProduct;
-		info->bmp_type = type;
-		strncpy(info->serial, serial, sizeof(info->serial));
-		strncpy(info->product, product, sizeof(info->product));
-		strncpy(info->manufacturer, manufacturer, sizeof(info->manufacturer));
-		if (cl_opts->opt_position && cl_opts->opt_position == found_debuggers + 1U) {
-			found_debuggers = 1;
-			break;
+static const probe_info_s *scan_for_devices(bmp_info_s *info)
+{
+	libusb_device **device_list;
+	libusb_device_descriptor_s device_descriptor;
+	probe_info_s *probe_list = NULL;
+
+	int result;
+	ssize_t cnt;
+	//
+	// If we are running on Windows the proprietary FTD2XX library is used
+	// to collect debugger information.
+	//
+#if defined(_WIN32) || defined(__CYGWIN__)
+	probe_list = process_ftdi_probe();
+	const bool skip_ftdi = probe_list != NULL;
+#else
+	const bool skip_ftdi = false;
+#endif
+
+	cnt = libusb_get_device_list(NULL, &device_list);
+	if (cnt > 0) {
+		// Parse the list of USB devices found
+
+		for (size_t deviceIndex = 0; device_list[deviceIndex]; ++deviceIndex) {
+			libusb_device *const device = device_list[deviceIndex];
+			result = libusb_get_device_descriptor(device, &device_descriptor);
+			if (result < 0) {
+				result = fprintf(stderr, "failed to get device descriptor");
+				return (probe_info_s *)NULL;
+			}
+			if (device_descriptor.idVendor != VENDOR_ID_FTDI || !skip_ftdi) {
+				if (!process_vid_pid_table_probe(&device_descriptor, device, &probe_list))
+					process_cmsis_interface_probe(&device_descriptor, device, &probe_list, info);
+			}
 		}
-		++found_debuggers;
+		// if (info->libusb_dev)
+		// 	libusb_ref_device(info->libusb_dev);
+		libusb_free_device_list(device_list, 1);
```

Comment:
This should be `libusb_free_device_list(device_list, count);` to be correct with the libusb API expectations and documentation

Reply:
Edit as requested.